icu4j/main/classes/core/src/com/ibm/icu/impl/breakiter/CjkBreakEngine.java

```diff
@@ -196,6 +236,25 @@ public int divideUpDictionaryRange(CharacterIterator inText, int startPos, int e
         if (bestSnlp[numCodePts] == kint32max) {
             t_boundary[numBreaks] = numCodePts;
             numBreaks++;
+        } else if (isPhraseBreaking) {
+            t_boundary[numBreaks] = numCodePts;
+            numBreaks++;
+            int prevIdx = numCodePts;
+            int codeUnitIdx = 0, length = 0;
+            for (int i = prev[numCodePts]; i > 0; i = prev[i]) {
+                codeUnitIdx = prenormstr.offsetByCodePoints(0, i);
+                length = prevIdx - i;
+                prevIdx = i;
+                String pattern = getPatternFromText(text, s, codeUnitIdx, length);
```

Comment:
FYI  FYI
Thinking about the code points vs. code units...
- The pre-increment next32() function is not super efficient, because it first has to do most of what current32() does just to skip the current code point.
- Therefore, your getPatternText() from a couple of days ago (taking a code unit length)  *might* be faster.
- It should also help to compute the codeUnitIdx of the first `i` before the loop, and then offsetByCodePoints() backwards from there in small increments of `i - prevIdx` (negative=backwards), tracking two variables like codeUnitIdx and prevCodeUnitIdx.
- More interesting: If you knew that your fSkipSet never contained any strings with supplementary characters (e.g., offsetByCodePoints() moves by more code units than the given code point delta), then maybe you could just detect if there are any supplementary code points in the "length" span, and if so, skip the get-pattern and the matching (fSkipSet.contains() would be false anyway).